subprojects/core-api/src/main/java/org/gradle/api/DomainObjectCollection.java

```diff
@@ -127,6 +139,9 @@ public interface DomainObjectCollection<T> extends Collection<T> {
 
     /**
      * Adds an {@code Action} to be executed when an object is removed from this collection.
+     * <p>
+     * This method will only be invoked for elements effectively added to the collection, and might cause their realization.
```

Comment:
"Might" cause realization? When? What the second sentence says?
What exactly does "observed" mean?

Reply:
But if you `add` a Provider, then this method will be called, even if the value was still lazy. And in order to call the method, the value will be realized.

Reply:
Perfectly fine, let's just say this then, because it's useful, whereas the "might" is not.